File: frontend/src/dominio/eventos/DomainEvent.ts
export type DomainEvent<TPayload = Record<string, unknown>> = {
  id: string
  type: string
  occurredAt: number
  payload: TPayload
}

export const createDomainEvent = <TPayload>(
  type: string,
  payload: TPayload,
): DomainEvent<TPayload> => ({
  id: crypto.randomUUID(),
  type,
  occurredAt: Date.now(),
  payload,
})

File: backend/internal/domain/events/domain_event.go
package events

import (
	"crypto/rand"
	"encoding/hex"
	"time"
)

type DomainEvent struct {
	ID         string `json:"id"`
	Type       string `json:"type"`
	OccurredAt int64  `json:"occurredAt"`
	Payload    any    `json:"payload"`
}

func New(eventType string, payload any) DomainEvent {
	return DomainEvent{
		ID:         newEventID(),
		Type:       eventType,
		OccurredAt: time.Now().UnixMilli(),
		Payload:    payload,
	}
}

func newEventID() string {
	buffer := make([]byte, 8)
	if _, err := rand.Read(buffer); err != nil {
		return "evt_fallback"
	}
	return "evt_" + hex.EncodeToString(buffer)
}
